tide/master/rest/FileSystemQuery.cpp

```diff
@@ -0,0 +1,114 @@
+/*    THIS  SOFTWARE IS PROVIDED  BY THE  UNIVERSITY OF  TEXAS AT    */
+/*    AUSTIN  ``AS IS''  AND ANY  EXPRESS OR  IMPLIED WARRANTIES,    */
+/*    INCLUDING, BUT  NOT LIMITED  TO, THE IMPLIED  WARRANTIES OF    */
+/*    MERCHANTABILITY  AND FITNESS FOR  A PARTICULAR  PURPOSE ARE    */
+/*    DISCLAIMED.  IN  NO EVENT SHALL THE UNIVERSITY  OF TEXAS AT    */
+/*    AUSTIN OR CONTRIBUTORS BE  LIABLE FOR ANY DIRECT, INDIRECT,    */
+/*    INCIDENTAL,  SPECIAL, EXEMPLARY,  OR  CONSEQUENTIAL DAMAGES    */
+/*    (INCLUDING, BUT  NOT LIMITED TO,  PROCUREMENT OF SUBSTITUTE    */
+/*    GOODS  OR  SERVICES; LOSS  OF  USE,  DATA,  OR PROFITS;  OR    */
+/*    BUSINESS INTERRUPTION) HOWEVER CAUSED  AND ON ANY THEORY OF    */
+/*    LIABILITY, WHETHER  IN CONTRACT, STRICT  LIABILITY, OR TORT    */
+/*    (INCLUDING NEGLIGENCE OR OTHERWISE)  ARISING IN ANY WAY OUT    */
+/*    OF  THE  USE OF  THIS  SOFTWARE,  EVEN  IF ADVISED  OF  THE    */
+/*    POSSIBILITY OF SUCH DAMAGE.                                    */
+/*                                                                   */
+/* The views and conclusions contained in the software and           */
+/* documentation are those of the authors and should not be          */
+/* interpreted as representing official policies, either expressed   */
+/* or implied, of Ecole polytechnique federale de Lausanne.          */
+/*********************************************************************/
+
+#include "FileSystemQuery.h"
+
+#include "log.h"
+
+#include <QDir>
+#include <QJsonArray>
+#include <QJsonDocument>
+#include <QJsonObject>
+#include <QUrl>
+#include <QUuid>
+
+namespace
+{
+QJsonObject _toJSONObject( const std::string& data )
+{
+    const QByteArray input = QByteArray(data.c_str(), data.size( ));
+    const auto doc = QJsonDocument::fromJson( input );
+    if( doc.isNull() || !doc.isObject( ))
+    {
+        put_flog( LOG_INFO, "Error parsing JSON string: '%s'", data.c_str( ));
+        return QJsonObject{};
+    }
+    return doc.object();
+}
+}
+
+FileSystemQuery::FileSystemQuery( zeroeq::http::Server& server,
+                                  const MasterConfiguration& config ) :
+    _server ( server )
+   ,_config ( config )
+{
+    _server.handlePUT( "tide/ls", [this]( const std::string& received )
+    { return _handleDirectoryList( received ); } );
+
+}
+
+bool FileSystemQuery::_handleDirectoryList( std::string payload )
+{
+    const auto obj = _toJSONObject( payload );
+    if( obj.empty( ))
+        return false;
+
+    QString path = obj["dir"].toString();
+    QString configDir = _config.getContentDir();
+    QString sessionDir = _config.getSessionsDir();
+
+    QString fullpath = configDir + "/" + path ;
+    QDir directory ( fullpath );
+
+    if(! ( fullpath.startsWith( configDir ) ||
+           fullpath.startsWith( sessionDir )))
+        return false;
+
+    if( directory.exists( ))
+    {
+        std::shared_ptr<FileSystemContent> fsEntry;
```

Comment:
only use shared_ptr if the ownership of the object is going to be shared with another class. In this case, I would even try to replace the _fsContentList with a std::map<std::string,FileSystemContent>. This would simplify deletion of events. Also, right now the contents are being replaced and re-registered every time, they could be just "updated" if they already exist.